lib/iris/util.py

```diff
@@ -2145,3 +2146,59 @@ def _strip_metadata_from_dims(cube, dims):
             reduced_cube.remove_cell_measure(cm)
 
     return reduced_cube
+
+
+def mask_cube_from_shapefile(cube, shape, minimum_weight=0.0, in_place=False):
+    """Takes a shape object and masks all points not touching it in a cube
+
+    Finds the overlap between the `shape` and the `cube` in 2D xy space and
+    masks out any cells with less % overlap with shape than set.
+    Default behaviour is to count any overlap between shape and cell as valid
+
+    Parameters
+    -----------
+
+    shape : Shapely.Geometry object
+        A single `shape` of the area to remain unmasked on the `cube`.
+        If it a line object of some kind then minimum_weight will be ignored,
+        because you cannot compare the area of a 1D line and 2D Cell
+    cube : :class:`~iris.cube.Cube` object
+        The `Cube` object to masked. Must be singular, rather than a `CubeList`
+    minimum_weight : float , default=0.0
+        A number between 0-1 describing what % of a cube cell area must
+        the shape overlap to include it.
+    in_place : bool, default=False
+        Whether to mask the `cube` in-place or return a newly masked `cube`.
+        Defaults to False.
+
+    Returns
+    --------
+
+    iris.Cube
+        A masked version of the input cube, if in_place is False
+
+
+    See Also
+    ---------
+
+    :func:`~iris.util.mask_cube'
+
+    Notes
+    -------
+    To mask a cube from a shapefile, both must first be on the same coordinate system.
+    Shapefiles are mostly on a lat/lon grid with a projection very similar to GeogCS
+    The shapefile is projected to the coord system of the cube, then each cell
+    is compared to the shapefile to determine overlap and populate a true/false array
```

Comment:
Upon investigating the cause of this is that the longitude coordinate for this cube is in the range 312...390, but my adjustment for 360-based coordinate ranges only fires if the coord_system is the default. Think we discussed this around the last demo of this, and determined there was not easy fix. Can put in a warning that something may be going wrong, might be a cleverer way of dealing with it via the modulus. It's an existing problem with ASCEND aswell.

Reply:
Mmmm I vaguely remember the conversation. If it's an existing problem with ASCEND, and it's a situation we can detect, then I would be keen to raise an `Error` for these situations (as opposed to trying to fix it right now). The result will not be what the user expects/needs, so I prefer error to warning, and the user is always free to use try-except if they really want their script to continue.

Reply:
It's not really something we can detect automatically in all cases. Basically, the coord_system of a cube doesn't define the range of longitudes used for a cube - whether it's -180-180, 0-360, or 180-540 (the case for this cube). For some coord systems (eg OSGB) the "x" coordinates get massive (103976.3-652897.98) so you can't just blindly assume on the x values alone.

This is a common situation from rotated pole grids that their meridian is around 360 rather than 0, so I can write a check for this for other RotatedGeogCS systems. But that doesn't mean this same problem won't crop up again in other coord systems.

Reply:
OK that does sound more like a case for a `Warning` than an `Error`. I'm assuming we can avoid issuing the `Warning` ALL the time, but can detect cases where there _might_ be a problem (e.g. only when using a non-default coord system)?

Reply:
Yea - Here's the one I'm adding 


```python
    elif target_system != DEFAULT_CS and cube.coord(x_name).points[-1] > 180:
        # this may lead to incorrect masking or not depending on projection type so warn user
        warnings.warn(
            """Cube has x-coordinates over 180E and a non-standard projection time.\n
              This may lead to incorrect masking. \n
             If the result is not as expected, you might want to transform the x coordinate points to -180 to 180  """,
            category=IrisUserWarning,
        )

Reply:
Talking to @nhsavage  this is a known issue in the CORDEX/CF community about definition of rotated poles and where their meridians are.

[https://github.com/SciTools/cartopy/issues/2091]

[https://github.com/SciTools/iris/issues/4978]